doc/design/udf_discovery_and_communication.rst

```diff
@@ -0,0 +1,63 @@
+UDF Discovery and Communication
+===============================
+
+===============
+Local Discovery
+===============
+
+*********
+Overview:
+*********
+
+- UDP Broadcast for the initial PingMessage with connection information for the receiving socket of the reliable network
+- After receiving PingMessage:
+
+  - Create sending socket for peer to its receiving socket of the reliable network
+  - Send a ReadyToReceiveMessage with our connection information for the receiving socket
+    over our sending socket for it to inform the peer that we are ready to receive from it
+
+- After receiving ReadyToReceiveMessage on our receiving socket:
+
+  - If not yet discovered, we create a sending socket for the peer to its receiving socket of the reliable network
+  - Send a ReadyToReceiveMessage with our connection information for the receiving socket
+    over our sending socket for it, in case we didn't get its PingMessage.
+  - If we didn't get a ReadyToReceiveMessage message after a certain time , yet, we send a AreYouReadyToReceiveMessage to a discovered peer
+    - This should prevent a stuck handshake if we should lose a ReadyToReceiveMessage for whatever reason
+
+- After receiving AreYouReadyToReceiveMessage:
+
+  - If not yet discovered, we create a sending socket for the peer to its receiving socket of the reliable network
+  - Send a ReadyToReceiveMessage with our connection information for the receiving socket
+    over our sending socket for it, in case we didn't get its PingMessage.
+
+.. image:: peer_communicator_state_machine.drawio.png
+
+********
+Details:
+********
+
+We seperated the Local Discovery into two components. The component LocalDiscovery implements
```

Comment:
You use the term Local Discovery for the whole communication process, but also for one component. This could be a bit confusing.

Reply:
Maybe we split into Discovery and Cluster Formation. Or rename the component into LocalDiscoveryStrategy. However, in general, the handshakes the PeerCommunicator does, are part of the Discovery, because only afterwards, it is guaranteed that all peers know each other. Need to think about it.